### ISSUE TITLE
2.0.0 release

### ISSUE THREAD

**Issue by blairconrad** [FakeItEasy/FakeItEasy#336]
**Ready** when all other issues on this milestone are **Done** and closed.

- [x] check integration package compatibility
   - [x] AutoFixture.AutoFakeItEasy
     - stable *incompatible* with 2.0.0-rc2, but [restricted to less than 2.0](https://github.com/AutoFixture/AutoFixture/pull/617#issuecomment-217612774)
   - [x] Machine.Fakes.FakeItEasy
     - stable *incompatible* with 2.0.0-rc1
     - pre-release compatible with  2.0.0-rc2
   - [x] Autofac.Extras.FakeItEasy
     - stable *incompatible* with 2.0.0-rc1 but [restricted to less than 2.0](https://github.com/autofac/Autofac.Extras.FakeItEasy/pull/1)
     - pre-release compatible with  2.0.0-rc2
- [x] run code analysis in VS in *Release* mode and address violations (send a regular PR which must be merged before continuing) - #336
- [x] release RC https://www.nuget.org/packages/FakeItEasy/2.0.0-rc2 (1st May 2016)
- [x] ensure the RC has been available for *some period of time* (box checked on 20th May, almost 3 weeks after release)
- [x] smoke test the analyzer package. We're not releasing it, so just download the nupkg and install from local disk
- [x] if necessary, change `VERSION_SUFFIX` on [CI Server](http://teamcity.codebetter.com/admin/editBuildParams.html?id=buildType:bt929)
      to appropriate "-beta123" or "" (for non-betas) value and initiate a build
- [x] check build
-  edit draft release in [GitHub UI](https://github.com/FakeItEasy/FakeItEasy/releases):
   - [x] complete release notes, mentioning non-owner contributors, if any
   - [x] attach nupkg
   - [x] publish the release
- [x] remove https://github.com/FakeItEasy/FakeItEasy/wiki/Changes-in-version-2.0
- [x] push NuGet package
- [x] de-list pre-release or superseded buggy NuGet packages if present
- [x] tweet, mentioning contributors and post link as comment here for easy retweeting ;-)
- [x] post tweet in [Gitter ](https://gitter.im/FakeItEasy/FakeItEasy)
- [x] post links to the GitHub release in each issue in this milestone, with thanks to contributors
- [x] use `rake set_version[new_version]` to 
   - create a new branch
   - change CommonAssemblyInfo.cs to expected minor version (of form _xx.yy.zz_)
   - push to origin
   - create PR to upstream master
- [x] use `rake create_milestone` to
   - create a new milestone for the next release
   - create new issue (like this one) for the next release, adding it to the new milestone
   - create a new draft GitHub Release 
- [x] close this milestone

**Comment by adamralph** [comment]
Awaiting an assessment of #156, #224 and #316 for inclusion in 2.0.

**Comment by blairconrad** [comment]
Things seem to be winding down nicely in the [2.0 milestone](https://github.com/FakeItEasy/FakeItEasy/milestones/2.0.0). There are 2 unclaimed issues (I think), and 3 in flight, 2 from a community member.

I am inclined to think about closing off 2.0. Does anyone have anything they think should go in that we've not planned in the milestone? I'm happy to leave non-breaking enhancements and bug fixes 'til later, but if there's something you're keen on, speak up.

Related to that, once we think we're feature complete, should we switch to an "-rc01" release from the current "-betaxyz" format?

**Comment by thomaslevesque** [comment]
> Does anyone have anything they think should go in that we've not planned in the milestone?

No, I think it's good. I have a few ideas for improvements, but it can wait until a future release.

> Related to that, once we think we're feature complete, should we switch to an "-rc01" release from the current "-betaxyz" format?

Well, typically, a beta is promoted to RC when it's thought to be stable. Will the release including the remaining issues be more stable than the previous beta? I don't know... DId we even receive any feedback on the beta versions? So far there has only been 77 downloads of beta-011, so it's hard to tell.

**Comment by blairconrad** [comment]
Oh, I'd always thought the RCs came in at "feature complete" time.
Really, none of the betas are any more or less stable than any of the actual production releases! We're just calling them betas because we don't promise not to change the API. And if we think that we have no more breaking changes coming, then I guess we're close to retiring the promise, so... maybe more stable?

**Comment by thomaslevesque** [comment]
> maybe more stable?

More stable in terms of API surface, anyway... OK, I guess we can make it a RC.

**Comment by adamralph** [comment]
For me, RC means "we've done all the work and we think we can release this". Ideally, it will just remain available for a period of time for the community to test it and then we can just release RTM without further commits (take a look at xunit 2.1).

The only reason to consider adding another feature to 2.0 now is if it is breaking, since we ought to limit the rate of breaking releases.

@thomaslevesque if you have more feature ideas, raise them now rather than later. You may not think they are breaking but perhaps someone may spot something.

Generally, I'd say any further breaking change would have to be hugely compelling to add it to 2.0 now. I'm very keen to get 2.0 out ASAP.

**Comment by adamralph** [comment]
> I have a few ideas for improvements, but it can wait until a future release

Once we are past 2.0, we are again at liberty to release every time a feature or bug fix hits master. Most of our 1.x releases only ever contained a handful, or sometimes even just one, feature or bug fix.

**Comment by blairconrad** [comment]
> Once we are past 2.0, we are again at liberty to release every time a feature or bug fix hits master. Most of our 1.x releases only ever contained a handful, or sometimes even just one, feature or bug fix.

And I'm itching for the day. I really enjoyed that "release almost exactly every 3 weeks" cadence we had going for a while.

**Comment by thomaslevesque** [comment]
> For me, RC means "we've done all the work and we think we can release this". Ideally, it will just remain available for a period of time for the community to test it and then we can just release RTM without further commits (take a look at xunit 2.1).

That's what I had in mind initially when @blairconrad suggested making it RC, but I don't have a strong opinion either way.

> @thomaslevesque if you have more feature ideas, raise them now rather than later. You may not think they are breaking but perhaps someone may spot something.

I already raised on here: https://github.com/FakeItEasy/FakeItEasy/issues/578
(probably not breaking)

**Comment by thomaslevesque** [comment]
Another idea that I had in mind was somewhat related to #578. I really like the "fixture injection" feature, but it would be nice if it could behave more like a real IoC container; so I thought we could provide a way to somehow plug an IoC container into the `Fake.InitializeFixture` mechanism. But I don't yet have a clear idea of how it would work, or even if we could do it without introducing a breaking change.

**Comment by blairconrad** [comment]
Perhaps this isn't the place to discuss it, but I wonder how much further we want to extend the "FakeItEasy as test dependency provider". The basic functionality is cool, but in the end are we just going to replicate [FakeItEasy](https://github.com/AutoFixture/AutoFixture/tree/master/Src/AutoFakeItEasy)?

And if so, that's not necessarily a bad thing. I just want to make sure we'd be doing it intentionally…

**Comment by thomaslevesque** [comment]
> Perhaps this isn't the place to discuss it, but I wonder how much further we want to extend the "FakeItEasy as test dependency provider". The basic functionality is cool, but in the end are we just going to replicate AutoFakeItEasy?
> 
> And if so, that's not necessarily a bad thing. I just want to make sure we'd be doing it intentionally…

I agree that it's not a core feature, but it's really convenient. AutoFakeItEasy is good, but it doesn't inject fakes into the test fixture itself. I'll create an issue for this when the idea is a bit clearer in my mind

**Comment by blairconrad** [comment]
:+1: Thanks!

**Comment by blairconrad** [comment]
I've taken a pass at the remaining issues, and searched my brain, and saw nothing else that might be breaking that we might want to put in 2.0.
Has anyone anything else, aside from PR #598?

If not, I'd suggest that we take the next build we get, test it against our big 3 NuGet consumers as listed way above, plus anything else someone would like to try, and if all goes well, brand that version the release candidate. What say you all?

**Comment by adamralph** [comment]
@blairconrad I'm :+1: to that. I had a play around with xbehave regarding #598 but I'm having a hard time figuring out how to do it. I suppose, for now we could abandon xbehave for the offending test. The problem is an xbehave problem, so we don't need to burden ourselves with it here at FIE.

**Comment by blairconrad** [comment]
>  we could abandon xbehave for the offending test. The problem is an xbehave problem, so we don't need to burden ourselves with it here at FIE.

Oh, just use naked XUnit for that spec?
Alternatively, we might be able to keep xbehave with a less than ideal step structure?

I'd originally thought to wait on #598, as it's unlikely to be breaking, but I think this is a better plan.

**Comment by blairconrad** [comment]
@adamralph I just realized we're talking about #598, but we might've meant #575?

**Comment by adamralph** [comment]
@blairconrad oops, yes, #575 is the one I was referring to.

The interesting thing there is that whilst the issue doesn't imply a breaking change (#484), the proposed implementation in #575 is a breaking change, as evident in the breaking of our xbehave-based test. It's arguable that all test frameworks should be behaving properly and flowing context, but there may be other subtleties involved in the switch to using logical call context - it's a horridly complex area. There's also the .NET 4.0 vs 4.5 consideration.

**Comment by blairconrad** [comment]
Gah. I love/hate it when you're right.
I'd almost put it in the "clients doing something stupid, so if we break it, it's their fault" category (hard on xBehave, I know, but the accidental logical thread context switching is perhaps not the best overall behaviour),  but I really _do_ enjoy not inconveniencing people. So let's do fix the test (by avoiding the problem - have you a preferred method?) and the Scopes' behaviour.

Can you expand on the the "4.0 vs 4.5 consideration"? I'm ever so slightly lost.

**Comment by adamralph** [comment]
> Can you expand on the the "4.0 vs 4.5 consideration"? I'm ever so slightly lost.

See https://github.com/FakeItEasy/FakeItEasy/pull/575#issuecomment-170486697

As to the actual concrete effect, I'm also ever so slightly lost. :smile: I've done my best to get up to speed with `ExecutionContext`, `SynchronizationContext`, etc. to try and solve the xbehave problem, but it's fiendishly complicated.

**Comment by blairconrad** [comment]
Too much braining for 7:15. Upshot sounds like the new Scope code may not be so good on 4.0?

**Comment by adamralph** [comment]
> Upshot sounds like the new Scope code may not be so good on 4.0?

Yep, that is one risk. There may be others.

**Comment by thomaslevesque** [comment]
> Upshot sounds like the new Scope code may not be so good on 4.0?

If I understand correctly, it means that on 4.0, changes to the logical call context from a called async method will affect the calling method as well, whereas in 4.5 it doesn't (because the context is copied before any modification)

Anyway, even for 4.5, I don't think it's a breaking change; the current behavior is broken, so I doubt anyone is relying on it.

**Comment by adamralph** [comment]
> Anyway, even for 4.5, I don't think it's a breaking change; the current behavior is broken, so I doubt anyone is relying on it.

If anyone is using fake scope in xbehave tests, or any other test framework which doesn't flow execution context, then it will break them.

**Comment by thomaslevesque** [comment]
> If anyone is using fake scope in xbehave tests, or any other test framework which doesn't flow execution context, then it will break them.

Of course, I wasn't thinking clearly ^^

**Comment by adamralph** [comment]
Let's move the fake scope/execution context conversation over to https://github.com/FakeItEasy/FakeItEasy/pull/575#issuecomment-181557226

**Comment by blairconrad** [comment]
I think we're missing the last documentation piece for the new ordered assertions API, and… that may be it. Shall we "check integration package compatibility" from the checklist above and release rc01?

**Comment by blairconrad** [comment]
I'm content to document the new ordered assertions API, but would also be happy if someone else did.

**Comment by blairconrad** [comment]
I checked AutoFixture and updated [Ordered Assertions](https://github.com/FakeItEasy/FakeItEasy/wiki/Ordered-assertions).

**Comment by thomaslevesque** [comment]
> I checked AutoFixture and updated Ordered Assertions

Looking good, thanks @blairconrad!

**Comment by adamralph** [comment]
Indeed @blairconrad , the wiki page looks great!

**Comment by blairconrad** [comment]
I dunno. It's just a page. Thanks.
:blush:

**Comment by blairconrad** [comment]
Machine.Fakes.FakeItEasy is not compatible with FakeItEasy 2.0, as I noted above. I've created an issue and offered to help "upgrade" it.
This should be sufficient for us to continue, no?

**Comment by blairconrad** [comment]
(Not release 2.0 yet. I just mean that we shouldn't stop the release for Machine.Fakes.FakeItEasy, should we.)

**Comment by blairconrad** [comment]
Autofac.Extras.FakeItEasy is slightly less compatible than Machine.Fakes.FakeItEasy is.

**Comment by adamralph** [comment]
I think we should 

> run code analysis in VS in Release mode and address violations (send a regular PR which must be merged before continuing)

and then push out an RC, publicise it, wait a week or so and then push out 2.0 (the stable release should be built from the same commit as the latest RC).

**Comment by adamralph** [comment]
BTW I don't think we should delay the release due to issues with Machine.Fakes.FakeItEasy or Autofac.Extras.FakeItEasy. I don't believe they are widely used anyway. The most important one is AutoFixture.AutoFakeItEasy and we already know that works.

**Comment by thomaslevesque** [comment]
I also don't think we should wait for libraries that depend on us to be ready.

**Comment by blairconrad** [comment]
shall we be bold? `-rc1`?

**Comment by blairconrad** [comment]
… it limits us to only _9_ attempts!

**Comment by thomaslevesque** [comment]
> shall we be bold? -rc1?

Yes!

> … it limits us to only 9 attempts!

Not really, we could have rc10, rc42... rc999 :wink:. But hopefully that won't be necessary.

**Comment by blairconrad** [comment]
But the sorting would be shot to heck, @thomaslevesque.
Also, building right now!

**Comment by blairconrad** [comment]
https://twitter.com/fakeiteasyfx/status/707538433699348480

**Comment by adamralph** [comment]
I checked off 

> run code analysis in VS in Release mode and address violations (send a regular PR which must be merged before continuing)

I added and checked off

> release RC

and I added

> wait some period of time

Presumably, if any bugs are found (and fixed), then we clear the check list and start again?

**Comment by blairconrad** [comment]
Thanks, @adamralph.

Yeah, depending on the nature of the bug, I would probably not uncheck, recheck, and check the three dependent packages, but the others for sure.

**Comment by adamralph** [comment]
:+1: makes sense

**Comment by adamralph** [comment]
Given that we've decided to put #637 into 2.0, I've added a checklist item for it and cleared the "code analysis" and "release RC" items.

Once we have the PR for #637, we can decide if we also want to re-check the integration packages.

**Comment by blairconrad** [comment]
I'm all for clearing the "code analysis" and "release RC", but amn't sure why having #637 in the milestone isn't enough. I don't mind, though.

**Comment by adamralph** [comment]
We have #626 in there too. I guess it's because these issues were raised after this was brought into progress.

We could replace them both by putting "close all issues on this milestone" as the first item.

Currently, we have that as the second to last item and I'm not sure why since, by that point, this issue should be the only open issue left on the milestone.

**Comment by adamralph** [comment]
So, all 2.0.0 issues are closed, so this is once again `ready`.

I propose we release 2.0.0-rc2 immediately.

**Comment by thomaslevesque** [comment]
> I propose we release 2.0.0-rc2 immediately.

:+1: 

Let me just merge your last PRs

**Comment by adamralph** [comment]
I've released 2.0.0-rc2.

Do we want to re-test the integration packages?

**Comment by thomaslevesque** [comment]
> Do we want to re-test the integration packages?

That would probably be better. How do we test them? Just get the code, update to FIE 2.0.0-rc and run the tests?

**Comment by adamralph** [comment]
Yep, that would be sufficient. I agree, we should re-test.

**Comment by adamralph** [comment]
(We probably should have done it before releasing the RC.) :worried:

**Comment by blairconrad** [comment]
@thomaslevesque, that's what I did.

**Comment by adamralph** [comment]
I've completed

> check integration package compatibility

so the only thing remaining now is to leave the RC available for some time and then we can proceed with the release.

**Comment by adamralph** [comment]
https://twitter.com/adamralph/status/727924010856660993

**Comment by adamralph** [comment]
I smoke tested RC2 of the analyzer package. All good.

I've built an RTM package - all good.

I've also removed "posting the tweet in JabbR". https://jabbr.net doesn't even work anymore, and we officially migrated to Gitter some time back. Other locations to consider posting would be https://jabbr.slack.com/messages/general/details/ (a Slack replacement for the ex-JabbR community) and https://xunit.slack.com/messages/general/details/. There's also the dubiously named https://aspnetcore.slack.com/messages/general/details/ (when you go to the site, the team is titled ".NET") but I'm unsure if this is supposed to be .NET core/vnext/whatever-it's-called specific. I've [asked the question](https://aspnetcore.slack.com/archives/general/p1463773745013626).

**Comment by blairconrad** [comment]
The balkanized rooms make me sad.

**Comment by adamralph** [comment]
@FakeItEasy/owners I've completed the release notes. Can you please review? https://github.com/FakeItEasy/FakeItEasy/releases

As discussed, the release is currently scheduled for Wednesday 25th May 14:00 UTC.

**Comment by adamralph** [comment]
Argh!!!

![image](https://cloud.githubusercontent.com/assets/677704/15442491/d6e5a6dc-1ee1-11e6-9a3b-1b187a873fad.png)

I'm tempted to just leave it, and use [`FormattableString.Invariant()`](https://github.com/config-r/config-r/blob/0e7ec17bc5b03bbcadbe182db3847673dedab8dc/src/ConfigR/Internal/ObjectExtensions.cs#L21) when we move to .NET 4.6.

**Comment by blairconrad** [comment]
@adamralph

1. I reviewed the release notes. Your changes looked good. I made  few tiny updates to fix (unrelated) grammar/formatting problems.
2. We're moving to .NET 4.6? Instead of or addition to 4.0?

**Comment by adamralph** [comment]
Sorry, I should have said "*if/*when we move to .NET 4.6".

I just realised: the analyzer depends on Visual Studio 2015, which is shipped with .NET 4.6. Given that the analyzer isn't referenced, does that mean it can target .NET 4.6 and still be installed in a .NET 4.5 project? If so, we can upgrade the analyzer to .NET 4.6 right now. I'll test it...

**Comment by adamralph** [comment]
Yep, it works as I described. I created a .NET 4.0 project, installed the analyzer (currently .NET 4.5) with no problems, and the analyzer did it's business. I'll upgrade the analyzer to 4.6 and send a PR.

(BTW - I used FakeItEasy 1.25.3 and the analyzer still worked. :smile:)

**Comment by adamralph** [comment]
It turned out to be non-trivial to upgrade to .NET 4.6. I couldn't switch the version in the portable options (it kept telling me it would stick to 4.5 because they "support the same subset of APIs" or something) and the switch to .NET 4.6 only isn't supported by the tooling. I'd have to either create a new project, or hack the .csproj.

Instead, I just switched to a `string.Concat()` (which is actually more efficient) and raised #717.

**Comment by thomaslevesque** [comment]
> It turned out to be non-trivial to upgrade to .NET 4.6. I couldn't switch the version in the portable options (it kept telling me it would stick to 4.5 because they "support the same subset of APIs" or something) and the switch to .NET 4.6 only isn't supported by the tooling. I'd have to either create a new project, or hack the .csproj.

Yes, unfortunately it's not that simple... I'm not sure analyzers really *have* to be PCLs; analyzers targeting only .NET 4.6 would probably work in VS2015. But they won't work in other environment, e.g. VS Code on Linux (I'm not sure analyzers currently work in VS Code, but I expect they will at some point). Perhaps we could target just .NET 4.6 and .NET Core...

**Comment by adamralph** [comment]
> Perhaps we could target just .NET 4.6 and .NET Core...

Not in the land of PCL. AFAIK it will insist on targeting .NET 4.5. In any case, PCL is being replaced by netstandard, so when that is finalised we'll target whichever netstandard gives us the platforms we care about which as you say, will probably be .NET Core and .NET 4.6. Also, when we get a netstandard version of Castle.Core then we can switch FIE itself to a netstandard.

**Comment by thomaslevesque** [comment]
> Not in the land of PCL. AFAIK it will insist on targeting .NET 4.5.

Not for me:

![screenshot](http://i.imgur.com/8L7miCD.png)

**Comment by adamralph** [comment]
Ah, OK, with the others unchecked it behaves differently. In that case, yes, this is an option if we do need to target 4.6 rather than 4.5. I don't mind if we do this now or later since it won't be a breaking change (I think).

**Comment by thomaslevesque** [comment]
> I don't mind if we do this now or later since it won't be a breaking change (I think).

I'm not sure about the consequences, and there's no hurry, so I'd rather leave it as it is for now.

**Comment by blairconrad** [comment]
Tweet!

https://twitter.com/fakeiteasyfx/status/735468685683478528

**Comment by blairconrad** [comment]
At long-suffering last!

**Comment by adamralph** [comment]
Let the minor release train begin... :shipit: :

**Comment by blairconrad** [comment]
:steam_locomotive:
